core/src/main/java/org/mule/runtime/core/api/construct/FlowBuilder.java

```diff
@@ -0,0 +1,148 @@
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.api.construct;
+
+import static org.apache.commons.lang.StringUtils.isNotEmpty;
+import static org.mule.runtime.api.util.Preconditions.checkArgument;
+import org.mule.runtime.api.exception.MuleException;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.exception.MessagingExceptionHandler;
+import org.mule.runtime.core.api.processor.Processor;
+import org.mule.runtime.core.api.processor.strategy.ProcessingStrategyFactory;
+import org.mule.runtime.core.api.source.MessageSource;
+import org.mule.runtime.core.internal.construct.DefaultFlow;
+
+import java.util.List;
+
+/**
+ * Creates instances of {@link Flow} with a default implementation
+ * 
+ * <p/>
+ * Builder instance can be configured using the methods that follow the builder pattern until the flow is built. After that
+ * point, builder methods will fail to update the builder state.
+ */
+public class FlowBuilder {
+
+  private final String name;
+  private final MuleContext muleContext;
+  private MessageSource messageSource;
+  private List<Processor> messageProcessors;
+  private MessagingExceptionHandler exceptionListener;
+  private ProcessingStrategyFactory processingStrategyFactory;
+  private DefaultFlow flow;
+
+  /**
+   * Creates a new builder
+   *
+   * @param name name of the flow to be created. Non empty.
+   * @param muleContext context where the flow will be registered. Non null.
+   */
+  public FlowBuilder(String name, MuleContext muleContext) {
+    checkArgument(isNotEmpty(name), "name cannot be empty");
+    checkArgument(muleContext != null, "muleContext cannot be null");
+
+    this.name = name;
+    this.muleContext = muleContext;
+  }
+
+  /**
+   * Configures the message source for the flow.
+   *
+   * @param messageSource message source to use. Non null.
+   * @return same builder instance.
+   */
+  public FlowBuilder withMessageSource(MessageSource messageSource) {
+    checkImmutable();
+    checkArgument(messageSource != null, "messageSource cannot be null");
+    this.messageSource = messageSource;
+
+    return this;
+  }
+
+  /**
+   * Configures the message processors to execute as part of flow.
+   *
+   * @param messageProcessors message processors to execute. Non null.
+   * @return same builder instance.
+   */
+  public FlowBuilder withMessageProcessors(List<Processor> messageProcessors) {
+    checkImmutable();
+    checkArgument(messageProcessors != null, "messageProcessors cannot be null");
+    this.messageProcessors = messageProcessors;
+
+    return this;
+  }
+
+  /**
+   * Configures the exception listener to manage exceptions thrown on the flow execution.
+   *
+   * @param exceptionListener exception listener to use on the flow. Non null.
+   * @return same builder instance
+   */
+  public FlowBuilder managingExceptionsWith(MessagingExceptionHandler exceptionListener) {
```

Comment:
`handleErrorsWith`?  I think handle is clearer than manage, also we'd said we'd start talking about errors rather than exceptions in Mule and seen in other work done for Mule 4 e.g. `OnErrorContinue`